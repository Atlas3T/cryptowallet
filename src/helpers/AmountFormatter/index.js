/**
 * @TODO James Konrad
 * I've mocked some exchange rates below for the amount() function to use
 * Would make sense to put them somewhere in the Vuex store
 */

import numeral from 'numeral';

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GBP',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GBP',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GBP',
      rate: 9999,
    },
  ],
};

export default class AmountFormatter {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  getFormatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${this.currency.symbol}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `- ${formatted}`;
      if (amount > 0) return `+ ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency.code);
    if (rate) return this.amount * rate;
    return this.amount;
  }
}
