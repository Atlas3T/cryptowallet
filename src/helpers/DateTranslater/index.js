import { date as dateUtil } from 'quasar';

/**
 * Returns a translated date/month.
 * @param  {Number} timestamp
 * @param  {String} format
 * @param  {Object} vm Vue component instance
 * @return {String}
 */
export default function dateTranslater(timestamp, format, vm) {
  return dateUtil.formatDate(timestamp, format, {
    dayNames: [
      'sunday',
      'monday',
      'tuesday',
      'wednesday',
      'thursday',
      'friday',
      'saturday',
    ].map((item) => { return vm.$t(item); }),
    monthNames: [
      'january',
      'february',
      'march',
      'april',
      'may',
      'june',
      'july',
      'august',
      'september',
      'october',
      'november',
      'december',
    ].map((item) => { return vm.$t(item); }),
  });
}
