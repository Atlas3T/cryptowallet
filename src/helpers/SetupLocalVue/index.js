import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
