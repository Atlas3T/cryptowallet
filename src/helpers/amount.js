/**
 * @TODO James Konrad
 * mocked data below
 */

import numeral from 'numeral';

const currencySymbols = {
  USD: '$',
  GPB: 'Â£',
};

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GPB',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GPB',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GPB',
      rate: 9999,
    },
  ],
};

export default class Amount {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  get formatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${currencySymbols[this.currency]}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `&#45; ${formatted}`;
      if (amount > 0) return `&#43; ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency);
    return this.amount * rate;
  }
}
