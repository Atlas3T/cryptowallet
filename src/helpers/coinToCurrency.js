const exchangeRates = {
  btc: [
    {
      currency: 'USD',
      value: 6431.14,
    },
    {
      currency: 'GPB',
      value: 4977.06,
    },
  ],
  eth: [
    {
      currency: 'USD',
      value: 201.76,
    },
    {
      currency: 'GPB',
      value: 156.14,
    },
  ],
  atlas: [
    {
      currency: 'USD',
      value: 22222,
    },
    {
      currency: 'GPB',
      value: 9999,
    },
  ],
};

/**
 * Calcuates the value of a coin in a given currency
 * @param  {Number} amount
 * @param  {String} coin
 * @param  {String} currency
 * @return {Number}
 */

export default function coinToCurrency(amount, coin, currency) {
  const { value } = exchangeRates[coin].find(item => item.currency === currency);
  return amount * value;
}
