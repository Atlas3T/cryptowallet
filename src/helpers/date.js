import { date as dateUtil } from 'quasar';

/**
 * Returns formatted and translated date string
 * @param  {Number} timestamp
 * @param  {String} format
 * @param  {Object} vm Vue component instance
 * @return {String}
 */
export default function date(timestamp, format, vm) {
  return dateUtil.formatDate(timestamp, format, {
    dayNames: [
      'sunday',
      'monday',
      'tuesday',
      'wednesday',
      'thursday',
      'friday',
      'saturday',
    ].map(item => vm.$t(item)),
    monthNames: [
      'january',
      'february',
      'march',
      'april',
      'may',
      'june',
      'july',
      'august',
      'september',
      'october',
      'november',
      'december',
    ].map(item => vm.$t(item)),
  });
}
