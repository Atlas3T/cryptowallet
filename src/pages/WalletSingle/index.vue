<template>
  <div v-if="wallet">
    <Transactions :wallet="wallet" />
  </div>
</template>

<script>
import Transactions from '@/components/Wallet/Transactions';
import { mapState } from 'vuex';
// import { refreshWallet } from '@/helpers';

export default {
  name: 'WalletSingle',

  components: {
    Transactions,
  },

  data() {
    return {
      checkForUpdates: null,
      interval: 15000,
      timeout: 300000,
      startTime: null,
      balanceChanged: false,
    };
  },

  computed: {
    ...mapState({
      id: (state) => { return state.route.params.id; },
      authenticatedAccount: (state) => { return state.settings.authenticatedAccount; },
      delay: (state) => { return state.settings.delay; },
    }),

    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },
  },

  async activated() {
    let fullRefresh = false;
    this.startTime = new Date().getTime();
    this.checkForUpdates = setInterval(() => {
      const time = new Date().getTime();
      if (time - this.startTime > this.timeout) {
        clearInterval(this.checkForUpdates);
      }
      if (this.balanceChanged && this.wallet.sdk !== 'Bitcoin') { fullRefresh = true; }
      return this.refresher(() => {}, fullRefresh);
    }, this.interval);

    setTimeout(async () => {
      // console.log(await checkBalance(this.wallet));
      await this.refresher(() => {});
    }, this.delay.short);
  },

  deactivated() {
    this.worker = null;
    this.balanceChanged = false;
    clearInterval(this.checkForUpdates);
  },

  mounted() {
    this.$root.$on('updateWalletSingle', async (done) => {
      try {
        await this.backEndService.loadCoinPriceData(this.wallet.symbol);
        await this.refresher(done);
      } catch (err) {
        this.errorHandler(err);
        done();
      }
    });
  },

  methods: {
    /**
     * Performs a wallet update
     */
    async refresher(done, fullRefresh = true) {
      const online = window ? window.navigator.onLine : navigator.connection === 'none';
      if (online) {
        this.balanceChanged = await this.$walletWorker.refreshWallet(
          this.wallet, fullRefresh,
        );
        setTimeout(() => {
          done();
        }, this.delay.normal);
      } else { done(); }
    },
  },
};
</script>

<style scoped>

</style>
