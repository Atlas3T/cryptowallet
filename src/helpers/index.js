import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
