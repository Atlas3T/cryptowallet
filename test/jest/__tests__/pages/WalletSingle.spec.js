import WalletSingle from '@/pages/WalletSingle';
import { shallowMount } from '@vue/test-utils';
import { localVue } from '@/helpers/SetupLocalVue';
// import RefreshWallet from '@/helpers';
import Vuex from 'vuex';

jest.mock('@helpers/RefreshWallet', () => {
  return {
    refreshWallet: jest.fn(),
  };
});

describe('WalletSingle.vue', () => {
  let store;
  let getters;
  let wrapper;
  let backEndService;
  let errorHandler;
  let coinSDKS;

  beforeEach(() => {
    coinSDKS = {
      Bitcoin: jest.fn(),
    };
    errorHandler = jest.fn();
    backEndService = {
      connect: jest.fn(),
      loadPriceFeed: jest.fn(),
    };
    // refreshWallet = jest.fn();
    getters = {
      'entities/wallet/find': () => { return () => { return { sdk: 'Bitcoin' }; }; },
    };
    store = new Vuex.Store({
      state: {
        route: { params: { id: 1 } },
        settings: {
          authenticatedAccount: 1,
          delay: {
            normal: 500,
          },
        },
      },
      getters,
    });
    wrapper = shallowMount(WalletSingle, {
      localVue,
      store,
      mocks: {
        backEndService,
        errorHandler,
        coinSDKS,
      },
    });
  });

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('calls the refreshWallet plugin on a updateWalletSingle event', (done) => {
    wrapper.vm.$nextTick(() => {
      wrapper.vm.$root.$emit('updateWalletSingle');
      done();
    });
  });
});
