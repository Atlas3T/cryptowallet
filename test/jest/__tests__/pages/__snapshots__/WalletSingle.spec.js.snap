// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletSingle.vue renders and matches snapshot 1`] = `<!---->`;
